src/Init/Prelude.lean

```diff
@@ -2435,8 +2444,14 @@ class GetElem (cont : Type u) (idx : Type v) (elem : outParam (Type w)) (dom : o
   * `arr[i]'h`: uses `h` to prove the side goal
   -/
   getElem (xs : cont) (i : idx) (h : dom xs i) : elem
+  /--
+  The `getElem? arr i` function (notation `arr[i]?`) gets the nth element of `arr`,
+  or `none` if the index is out of bounds.
+  -/
+  getElem? (xs : cont) (i : idx) [Decidable (dom xs i)] : Option elem :=
```

Comment:
```suggestion
  getElem? (xs : cont) (i : idx) : Option elem :=
```
I wonder if it makes sense to bundle decidability here.

Reply:
I would love to, but what would the default implementation be in that case?

Reply:
```lean
local macro "get_elem?_default" : tactic =>
  `(exact dite .....)

-- ...
  getElem? (xs : cont) (i : idx) : Option elem := by get_elem?_default
```

Unfortunately Sebastian has built in all kinds of presumably helpful checks, so you can't just write `:= by exact ...` and need to do this dance with the separate macro.

Reply:
Ironically, we can't use that syntax yet, because this is in `Init.Prelude` and `macro` isn't legal. I think we can just move `GetElem` out of the prelude though...

Reply:
@gebner Well, even after the big refactor necessary to get GetElem out of Init.Prelude, it still doesn't work. Here's a MWE:
```lean
local macro "get_elem?_default"  : tactic =>
  `(trace_state; exact fun xs i => if h : _ then some ($(Lean.mkIdent `getElemₓ) xs i h) else none)

class GetElemₓ (cont : Type u) (idx : Type v) (elem : outParam (Type w)) (dom : outParam (cont → idx → Prop)) where
  getElemₓ (xs : cont) (i : idx) (h : dom xs i) : elem
  getElemₓ? : (xs : cont) → (i : idx) → Option elem := by get_elem?_default 

instance : GetElemₓ (Array α) USize α fun xs i => i.toNat < xs.size where
  getElemₓ xs i h := xs.uget i h
```
You can see in the printout that the auto param exists in the target local context (containing only `α`), which in particular lacks the `getElem` local variable that is available in regular default arguments. So there is no way for the auto param to know what was passed as the `getElem` argument to the structure.

Reply:
:sob: